### ISSUE TITLE
Auth problems with Apache reverse proxy

### ISSUE THREAD

**Issue by ramuyk** [shellhub-io/shellhub#2389]
### Description

I'm configuring ShellHub behind a reverse proxy with Apache, I've successfully installed ShellHub locally and I also managed to configure it with SSL using Apache. My Apache configuration is the following:

```
<VirtualHost *:80>

    ServerName mypage.example.com

    RewriteEngine On
    RewriteCond %{HTTPS} off 
    RewriteRule ^ https://%{SERVER_NAME}%{REQUEST_URI} [END,NE,R=permanent]

</VirtualHost>

<IfModule mod_ssl.c>
<VirtualHost *:443>

    ServerName mypage.example.com

    ProxyPreserveHost On

    RewriteEngine On
    RewriteCond %{HTTP:Upgrade}   =websocket    [NC]
    RewriteRule /(.*) ws://10.15.0.5:8088/$1 [P,L]

    ProxyPass / http://10.15.0.5:8088/
    ProxyPassReverse / http://10.15.0.5:8088/
    
    CustomLog ${APACHE_LOG_DIR}/mypage.example.log combined

    # SSL Certificates [HTTPS]
    SSLCertificateFile /etc/letsencrypt/live/mypage.example.com/fullchain.pem
    SSLCertificateKeyFile /etc/letsencrypt/live/mypage.example.com/privkey.pem
    Include /etc/letsencrypt/options-ssl-apache.conf

</VirtualHost>
</IfModule>
```

The problem is that when I'm configuring a device on ShellHub the device never sends the connection notification on my published https page. Looking at the Apache logs I see the following line:

```
10.15.1.1 - - [24/Jan/2023:19:58:05 -0300] "GET /api/devices/auth HTTP/1.1" 404 180 "http://mypage.example.com:80/api/devices/auth" "go-resty/2.7.0 (https://github.com/go-resty/resty)"
```

I'm not sure but my assumption is that the ShellHub agent on the device is forcing the request to arrive on port 80 with `http://mypage.example.com:80/api/devices/auth`.  I've realized that if I connect my device to ShellHub on the local network first and then I use my published page (mypage.example.com) everything works perfectly fine. But if I try to register new devices with my published page I never receive the notification that a device is trying to connect.

![2023-01-24_20-12](https://user-images.githubusercontent.com/44415096/214449168-5d5436fb-1535-4b4a-89e8-85ffd3e44ad2.png)

Is that a bug with the ShellHub agent?  Is there any way to make the `http://mypage.example.com:80/api/devices/auth` request successful while being behind a reverse proxy with Apache (considering I'm forcing it to be on the port 443)?

### Edition

Community

### Version

current master branch

**Comment by gustavosbarreto** [comment]
I tested your configuration and it worked perfectly.

Are you sure you installed the agent from the published page (mypage.example.com) instead of from 10.15.0.5?

**Comment by ramuyk** [comment]
Yeah, I'm sure of it

Actually, if I try to access the page http://10.15.0.5:8088 it works for registering the device:

![2023-01-25_13-58](https://user-images.githubusercontent.com/44415096/214629058-89f9ac52-bbf2-40da-ac4e-5e49d73e0597.png)

After registering the device I can control it from https://mypage.example.com ... The problem is when I try to register the device directly from mypage.example.com. Apparently, the ShellHub agent on my device is continuously trying to authenticate on mypage.example.com unsuccessfully:

![2023-01-25_14-03](https://user-images.githubusercontent.com/44415096/214630656-e586cc2e-36d6-447e-9eb3-65e0152177c7.png)

I'm not sure if I'm missing any details on my `.env.override` file, I only modified these three variables while building the containers:

```
SHELLHUB_HTTP_PORT=8088
SHELLHUB_HTTPS_PORT=8443
SHELLHUB_SSH_PORT=8022
```

To sum up, ShellHub works as long as I register the device for the first time in the same network where the server is (in my case http://10.15.0.5:8088)... But if the device doesn't have access to this network and I try to register it from mypage.example.com it doesn't work because the ShellHub agent fails to make the authentication on ShellHub.

**Comment by ramuyk** [comment]
I ended up finding a temporary solution, I was checking the file that builds the docker agent on the device:

![Uploading 2023-01-25_14-25.png…]()

If I change `http` to `https` it works fine. So a workaround for me is using a command like:

```
curl -sSf "https://mypage.example.com/install.sh?tenant_id=00000000-0000-0000-0000-000000000000" | sed 's/http/https/g' | sh
```

Even though it works now, for me it looks like a bug on the ShellHub agent. For some reason, it tries to force the connection on `http` by using `:80` at the end of the URL address...

**Comment by gustavosbarreto** [comment]
> If I change `http` to `https` it works fine. So a workaround for me is registering the device with a command like:
> 
> ```
> curl -sSf "https://mypage.example.com/install.sh?tenant_id=00000000-0000-0000-0000-000000000000" | sed 's/http/https/g' | sh
> ```

ShellHub install scripts needs the `X-Forwarded-Proto` header in order to define the protocol for the `SHELLHUB_SERVER_ADDRESS` variable. You can add this to your https virtualhost block in apache:

```
RequestHeader set X-Forwarded-Proto "https"
```

**Comment by ramuyk** [comment]
Thanks! Adding this line to my configuration completely solved the issue.
